src/fluree/db/json_ld/shacl.cljc

```diff
@@ -550,57 +597,65 @@
-                                                                               register-datatype property-shape)
-
-                                                                    (:node-kind property-shape)
-                                                                    (update-in [:datatype (:path property-shape)]
-                                                                               register-nodetype property-shape)
-
-                                                                    (:class property-shape)
-                                                                    (update-in [:datatype (:path property-shape)]
-                                                                               register-class class-iris))]
 
+                                             tagged-path (<? (resolve-path-types db path))
+
+                                             property-shape* (cond-> (assoc property-shape :path tagged-path)
+                                                               (:pattern property-shape) (assoc :pattern (build-pattern property-shape)))
+
+                                             p-shape-key (first (first tagged-path))
+                                             target-key  (first (peek tagged-path))
+                                             p-shapes*   (conj p-shapes property-shape*)
+                                             ;; elevate following conditions to top-level custom keys to optimize validations
+                                             ;; when processing txs
+
+                                             class-iris (when-let [class-sids (:class property-shape)]
+                                                          (loop [[csid & csids] class-sids
+                                                                 ciris          []]
+                                                            (let [ciri       (->> [csid const/iri-id]
+                                                                                  (query-range/index-range db :spot =)
+                                                                                  <?
+                                                                                  first
+                                                                                  flake/o)
+                                                                  next-ciris (conj ciris ciri)]
+                                                              (if (seq csids)
+                                                                (recur csids next-ciris)
+                                                                next-ciris))))
+
+                                             shape* (cond-> (update shape :properties
+                                                                    (fnil conj #{}) target-key)
+                                                      ;; hopefully obsolete
```

Comment:
Could you elaborate? You understand this comment now. Will you understand the same comment next year?

Reply:
Good catch, it is obsolete. I forgot to remove the comment and the condition.